bin/propolis-standalone/src/snapshot.rs

```diff
@@ -81,59 +93,78 @@ pub(crate) async fn save(
     let inv = guard.inventory();
 
     info!(log, "Serializing global VM state");
-    let global_state = {
-        let output =
+    {
+        let global_config =
             export_global(&hdl).context("Failed to export global VM state")?;
-        serde_json::to_vec(&output)?
-    };
+        let global_bytes = serde_json::to_vec(&global_config)?;
+        header.set_size(global_bytes.len() as u64);
+        header.set_cksum();
+        builder.append_data(&mut header, "global.json", &global_bytes[..])?;
+    }
+
+    // Add directories for devices and memory
+    {
+        let mut dir_header = tar::Header::new_gnu();
+        dir_header.set_entry_type(tar::EntryType::Directory);
+        dir_header.set_mode(0o555);
+        dir_header.set_mtime(now.as_secs());
+        dir_header.set_size(0);
+        dir_header.set_cksum();
+
+        builder.append_dir(&mut dir_header, format!("{DEVICE_DIR}/"))?;
+        builder.append_dir(&mut dir_header, format!("{MEMORY_DIR}/"))?;
```

Comment:
Can skip creating format strings
```suggestion
        builder.append_dir(&mut dir_header, DEVICE_DIR)?;
        builder.append_dir(&mut dir_header, MEMORY_DIR)?;
```